File: aws-sdk-swift/codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/plugins/DefaultAWSClientPlugin.kt
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */

package software.amazon.smithy.swift.codegen.integration.plugins

import software.amazon.smithy.aws.swift.codegen.swiftmodules.AWSClientRuntimeTypes
import software.amazon.smithy.codegen.core.Symbol
import software.amazon.smithy.swift.codegen.SwiftWriter
import software.amazon.smithy.swift.codegen.integration.Plugin

class DefaultAWSClientPlugin : Plugin {
    override val className: Symbol = AWSClientRuntimeTypes.Core.DefaultAWSClientPlugin

    override val isDefault: Boolean
        get() = true

    override fun customInitialization(writer: SwiftWriter) {
        writer.writeInline("\$N(clientName: self.clientName)", className)
    }
}

File: aws-sdk-swift/Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Plugins/DefaultAWSClientPlugin.swift
//
// Copyright Amazon.com Inc. or its affiliates.
// All Rights Reserved.
//
// SPDX-License-Identifier: Apache-2.0
//

import ClientRuntime

public class DefaultAWSClientPlugin: Plugin {
    private var clientName: String

    public init(clientName: String) {
        self.clientName = clientName
    }

    public func configureClient(clientConfiguration: ClientConfiguration) throws {
        if var config = clientConfiguration as? (DefaultClientConfiguration
            & AWSDefaultClientConfiguration
            & AWSRegionClientConfiguration) {
            config.retryStrategyOptions = try AWSClientConfigDefaultsProvider.retryStrategyOptions(
                config.awsRetryMode,
                config.maxAttempts
            )
            config.awsCredentialIdentityResolver = try AWSClientConfigDefaultsProvider.awsCredentialIdentityResolver()
        }
    }
}
